src/index.js

```diff
@@ -53,45 +54,50 @@ function createBabelPluginFactory(customCallback = returnObject) {
 			externalHelpersWhitelist,
 			include,
 			runtimeHelpers,
+			transformGenerated,
 			...babelOptions
 		} = unpackOptions(pluginOptions);
 
+		if (transformGenerated) {
+			return {
+				name: 'babel',
+
+				renderStart() {
+					if (pluginOptions.include || pluginOptions.exclude || pluginOptions.extensions) {
+						warnOnce(
+							this,
+							'The "include", "exclude" and "extensions" options are ignored when using the "transformGenerated" option.',
+						);
+					}
+				},
+
+				renderChunk(code) {
+					return transformCode(code, babelOptions, overrides, customOptions, this);
```

Comment:
We probably want to automatically wrap this in an IIFE, since Babel's transforms generally assume that their output will be a module. They don't worry about the potential for variable name collisions when things are hoisted to the top-level scope of a file, for instance.

Reply:
That depends on what people do but I am sure that for the majority of users, this is not desired as it will just add indirection. This is only interesting if your target is iife anyway and you want to use script tags. There will be no further transformations or hoisting after Babel does its thing.

Reply:
Ah right, I forgot that Rollup has those modes, I haven't actually used it much, just trying to provide insight from the side of Babel. Is `renderChunk` here guaranteed to run before Rollup applies it's UMD or IIFE `format` wrappers, so they'd wrap around any new top-level variables Babel creates?

Reply:
No, it is guaranteed to ran AFTER Rollup applies its transformations, so your comment is valid, but it would definitely not apply when targeting esm or cjs. Thus I wonder if we should address it at all, at least for the initial release.

Reply:
Could we have the plugin throw if the format isn't `cjs` or `esm`? My concern is that it's not a problem you're likely to notice until it starts causing really annoying problems that may not be obvious to users of the plugin.

Reply:
Throw would mean it is completely unusable even though it produced valid output, that seems like a bad choice for users. We could warn definitely. As I understand, it should theoretically be possible to set the output of Rollup to `esm` and then let `@babel/env` handle the conversion to other formats like `iife`, `amd` or `system`, is that correct?

Reply:
Yes, transforming esm with Babel should work.

Reply:
I really think throwing by default is important. For example, if a user uses a tagged template literal, they'll end up with a `_templateObject` global on their page. If anything else on the page happened to create that first, maybe because they loaded an advertising snippet or something that was also built with this plugin and happened to use a template literal, their site would then fail to load because it would be a duplicate declaration exception.

I would be absolutely terrified to have a application that declared anything in the global scope. It's super dangerous. If we don't throw by default, it would be the default behavior of this plugin.

Reply:
True, but it will not break anything for most cases. That is why I would go for a warning instead of an error. Warnings can easily be turned into errors by users if they wish to, but otherwise I see this as needlessly crippling functionality. I see this as similar to how Rollup handles `eval`–you can create a myriad of bugs, but we still allow it and instead go for a warning with a documentation link instead. I think we should do the same here. If you know what you are doing, this would allow for a much broader range of transformations to be applied via Babel in more situations.

Reply:
To take a step back, what exactly is the concern with adding an IIFE in the non-cjs/esm case? The only reason we're even arguing about error vs warn is because you were concerned about that, but an extra iife isn't going to dramatically affect anything, and guarantees this isn't an issue either way. 

If we landed this as-is, this plugin basically invalidate's Rollup's entire `format: "iife"` option by making the output no longer an IIFE, which is entirely counter to user expectations.

> Warnings can easily be turned into errors by users if they wish to

I would argue that no-one will ever do that, so it might as well be silent.

> I see this as needlessly crippling functionality

I really strongly disagree with that. It's not crippling, it's asking Babel to output code that is known to be unsafe to execute in a general context and specifically breaks the guarantees provided by Rollup's `format`.

Reply:
> To take a step back, what exactly is the concern with adding an IIFE in the non-cjs/esm case

1. I do not believe Rollup's wrappers will work properly when wrapped in another IIFE, e.g. a `var` will no longer create a global variable.
2. Rollup is chosen for minimal output to achieve a goal. Having an IIFE inside an IIFE is not minimal.

> If we landed this as-is, this plugin basically invalidate's Rollup's entire format: "iife" option by making the output no longer an IIFE

My limited experience with Babel is that without plugins, Babel does zero transformations, so what you said is obviously wrong. This still holds true if I hand-pick my plugins to just do certain transformations. For instance I might want to replace certain constructs with different ones. Or I actually want to modify the wrapper code, and for some reason, I am fluent with Babel so it is my tool of choice. In that case, I could happily live with a warning, but if the developers throw an error instead I would ask "Why on earth!? This is not my issue you are trying to fix!". To my experience, people usually notice warnings, and if there are no issues, what is the problem? I really do not understand why you are obsessed that it has to be an error.

 >I really strongly disagree with that. It's not crippling, it's asking Babel to output code that is known to be unsafe to execute in a general context
It is not known to be unsafe, this depends, as I said, entirely on the transforms you apply. I see it as a big improvement to allow arbitrary transforms because Babel is such a wide-spread AST transformation engine that will enable people to build more powerful stuff on Rollup. I guess for this particular topic, we will remain in disagreement.

Reply:
All of this assumes `format` is not `esm`/`cjs`:

> I do not believe Rollup's wrappers will work properly when wrapped in another IIFE, e.g. a var will no longer create a global variable.

Fair enough, I wasn't sure of Rollup created globals via `var` or undeclared assignment. This is another example where, since many of Babel's plugins assume that a file is a module, they would treat it as reasonable to rename that `var` expecting there to be no external behavioral changes though.

> My limited experience with Babel is that without plugins, Babel does zero transformations, so what you said is obviously wrong. 

That seems a bit harsh. Your own examples in the PR description and README talk about running `@preset/env` on the entire bundle, for instance, and make to recommendation against that in the `iife`/`umd` case. 99% of users will never stop to think of the specifics of the plugins they are running. Any plugin in the future could also change an implementation detail only to suddenly break a user of this patch's bundle by suddenly creating what it expected to be a module-level variable but is actually a global that conflicts with something somewhere.

> This still holds true if I hand-pick my plugins to just do certain transformations.

How are we expecting users to know that plugins are safe to hand-pick? I'd consider anything that creates a global helper function be be an immediate danger to use, but we don't document that and we could always add a helper in the future. I'd consider any `@babel/*` plugin to be unsafe to use on a global script without risk of future breakage.

> For instance I might want to replace certain constructs with different ones. Or I actually want to modify the wrapper code, and for some reason, I am fluent with Babel so it is my tool of choice. In that case, I could happily live with a warning, but if the developers throw an error instead I would ask "Why on earth!? This is not my issue you are trying to fix!". 

If you are someone who defines is fluent in writing a custom Babel transformation, you probably are aware of the pitfalls of transforming a global script, and you'd absolutely have no problem passing an extra option to opt-into global IIFE/UMD processing.

If you're _not_ someone who knows anything about Babel, but sees that this plugin and tries to give it a show, you're honestly saying that you think a warning is better than a clear error saying `Using this plugin to transform a global script with Babel's standard plugins is dangerous. If you are hoping to write your own plugin transform for Rollup's output, pass 'forceEnable:true' opt out of this verification, otherwise we recommend using the per-file transforms.`?

> To my experience, people usually notice warnings, and if there are no issues, what is the problem? I really do not understand why you are obsessed that it has to be an error.

My concern is specifically the "if there are no issues" mentioned in your comment. Running Babel on a full script specifically promotes an approach to Babel complication that is prone to _future_ unexpected issues, even if there are not _present_ issues. That means that someone can easily set it up, do what they need to do, and then get bitten unexpectedly and frustratingly later on. And when they do, it's not particularly satisfying to say "well you ignored that warning", whereas an error makes it clearer.

> It is not known to be unsafe, this depends, as I said, entirely on the transforms you apply. I see it as a big improvement to allow arbitrary transforms because Babel is such a wide-spread AST transformation engine that will enable people to build more powerful stuff on Rollup. I guess for this particular topic, we will remain in disagreement.

Again, your own examples right in the PR and README, have examples of using `preset-env` on the bundle as a whole. That's absolutely unsafe. What if another bundle loads on the page that is built with a different version of Babel and all of the now-global helper functions are overwritten with ones meant for another Babel version? Maybe your company has a general vendor bundle that gets loaded on every page and then once day it upgrades to Babel 8 but apps are still compiled with Babel 7 or something.

I'm not saying this feature isn't useful, I'm saying easily-misused by default is a bad default. Your `eval` example earlier I'd put at a much lower grade annoyance because it's going to be obvious that it is broken when you actually run your code. In this case of this, it'll seem to work until suddenly some random third-party script takes down your whole website because it injected a different version of one of Babel's helpers or something.

Look, if at the end of the day you'd still rather the warning, it's not the end of the world. That said, it's important to then narrow down exactly what the warning says and what plugins people would actually be using on an overall script. I'd be tempted to disable `babel.config.js` loading by default in this case  to further make it clear that any plugins used should be hand-picked for strictly Rollup's final bundle transformation. Any core `@babel/*` plugin should be assumed to potential inject new globals that could then conflict with other page scripts and thus break your app, so we'd want to very clearly say that users shouldn't run Babel's plugins unless they really know what they are doing.

Reply:
> I'd be tempted to disable babel.config.js loading by default in this case to further make it clear that any plugins used should be hand-picked for strictly Rollup's final bundle transformation.

I'm wondering if this shouldn't be the case there - I see a benefit of loading automatically a unified config, but this use case here is very specific and I'm a little bit afraid that for this case applying unified config might also break stuff badly. I'm on a fence here a little bit - hard to tell what are going to be users' expectations. Arguably a mistake should be fairly quickly noticed and we have `configFile: false` to opt-out of it - so maybe it's OK to keep it how it is now.

Reply:
I'd be fine with it being opt-in, but I don't feel super strongly about it. Given that this isn't really a mode we've had before, it's mostly new ground anyway, so perhaps being cautious about defaults is good?